Lecture 15: Statistical and Algorithmic Foundations of Deep Learning (#18)
sailinglab/pgm-spring-2019: _posts/2019-03-06-lecture-15.md

$$

Thus, we can find the gradient of this loss function with respect to a weight $w_j$:

$$
\begin{aligned}
\frac {\partial E_D [\bar{w}]}{\partial w_j} &= \frac {\partial}{\partial w_j} \frac{1}{2} \sum_d (t_d - o_d)^2 \\
&= \frac{1}{2} \sum_d 2 (t_d - o_d) \frac {\partial}{\partial w_j} (t_d - o_d) \\
&= \sum_d (t_d - o_d) \frac {- \partial o_d }{\partial w_j} \\
&= - \sum_d (t_d - o_d) \frac { \partial o_d }{\partial net_d} \frac { \partial net_d }{\partial w_i} \\
&= - \sum_d (t_d - o_d) o_d (1-o_d) x_d^i \\
\end{aligned}
$$

So, $\nabla E_d [\bar{w}] = - (t_d - o_d) o_d (1-o_d) \bar{x_d}$. Thus, the incremental perception learning algorithm does the following (with a learning rate of $\eta$):

For each training example $d$ in $D$:
1. compute gradient $\nabla E_d [\bar{w}]$
2. $\bar{w} = \bar{w} - \eta \nabla E_d [\bar{w}]$

This can also be done in batch mode, by averaging gradients across multiple examples.

### Backpropagation using a computational graph

This learning algorithm gets slightly more complicated when we introduce hidden units, circled in yellow below:

<figure id="NN">
      <img src="{{ 'assets/img/notes/lecture-15/NN.png' | relative_url }}" />
</figure>

> There is no direct target for these hidden units. So, we train the weights in these layers using backpropagation, or "reverse-mode differentiation". First, we can build a computaitonal graph, where $x$ is the input and $f(x)$ is the output, and nodes 2, 3, and 4 are the intermediate computations:

Reviewer: what does "target" here mean?

Reviewer 2: ```suggestion
Unlike neurons in the final layer, hidden units do not correspond to a particular class.  We train the weights in these layers using backpropagation, or "reverse-mode differentiation". First, we can build a computaitonal graph, where $x$ is the input and $f(x)$ is the output, and nodes 2, 3, and 4 are the intermediate computations:
```

<figure id="graph">
      <img src="{{ 'assets/img/notes/lecture-15/graph.png' | relative_url }}" />
</figure>

By applying the chain rule, we can derive:

$$

> \frac {\partial f_n}{\partial x} = \sum_{i_1\in \pi(n)} \frac {\partial f_n}{\partial f_{i_1}} \frac {\partial f_{i_1}}{\partial x} = \sum_{i_1\in \pi(n)} \frac {\partial f_n}{\partial f_{i_1}} \sum_{i_2\in \pi(i_1)} \frac {\partial f_{i_1}}{\partial f_{i_2}} \frac {\partial f_{i_2}}{\partial x}  = \cdots

Reviewer: should this be `\prod` instead of `\sum`?

Reviewer 2: I believe this is \sum. Refer to slide 14: https://www.dropbox.com/s/w6zpjmy5xk0qtmp/lecture15-NN.pdf?dl=0

$$

This algorithm of recursively calling the chain rule is called backpropagation, and we can use it to train most deep networks, even though we don't have a target for the hidden units.

### Modern building blocks of deep networks

Most deep networks are composed of many simple building blocks:

1. Layers, specifying the architecture and type of connections
  * Fully connected
  * Convolutional and pooling
  * Recurrent

>   * ResNets

Reviewer: should be residual layer instead of "ResNets"

Author: Yes.

```suggestion
  * Residual Layers
```

  * etc.
3. Loss functions, which we are trying to minimize (can have many in one network)
  * Cross-entropy loss
  * Mean squared error
  * etc.

### Feature learning

Often one important goal of deep networks is to learn a good representation of the input. Deep networks are often good at doing this, and can attempt to get a "disentangled" representation, i.e. amenable to linear separation, as shown below:

<figure id="features">
      <img src="{{ 'assets/img/notes/lecture-15/features.png' | relative_url }}" />
</figure>

Because DNNs have this property, it is often possible to *transfer* representations and use pretrained deep neural networks as feature extractors for new tasks. In these cases, the activations of intermediate layer $h_j$ in model $M$ (trained on task $A$) are used to generate representations for a different task $B$. Because these representations might be linearly separable, it typically suffices to train a linear model, or shallow neural network on the featurized data. The most common example of this is ImageNet, and models trained on ImageNet (i.e. vgg16) are typically good feature extractors for other vision based tasks. 

> ### Graphical Models vs Deep Networks

Reviewer: vs -> vs.

Reviewer 2: ```suggestion
### Graphical Models vs. Deep Networks
```

> While the two paradigms appear to share the same the graphical structure on the surface, juxtaposing their properties point out subtle differences between them. 

Reviewer: delete the extra "the"

Reviewer 2: ```suggestion
While the two paradigms appear to share the same graphical structure on the surface, juxtaposing their properties point out subtle differences between them. 
```

| **Structural utility**     | Graphs help monitor theoretical and empirical behaviour of inference | Networks facilitate the organization of computational operations |

**Tradeoffs**: As with any modeling strategy, there are inherent tradeoffs between graphical models and neural networks. Graphical models allow us to directly encode prior knowledge about the world into the structure of the model (i.e. latent states), enabling us to enforce some weak notions of causality. However, this requires domain expertise, and assumes that the structure enforced on the model is the most natural/well suited for the task. In contrast, neural networks learn their own latent representations, by minimizing loss in an end-to-end manner. As a result, they may be seen as learning a more 'efficient' representation. However, this approach can require signicant amounts of data, which may not always be available.

Thus, we see that the graphs in a graphical model represent the model itself whereas it just represents computation in the case of a deep network.

Some neural networks can be completely understood as a conventional graphical model. A few examples are examined below - 

> ## 1.Restricted Boltzmann Machines

Reviewer: make this subsubsubsections or paragraphs (instead of ##)?

Author: I think it makes more sense to make these ## rather than paragraphs.

<figure id="RBM">
      <img src="{{ 'assets/img/notes/lecture-15/RBM.png' | relative_url }}" />
</figure>

> **Overview**: RBMs are represented by undirected bipartite graphs, with one "visible" layer a second "hidden" layer. There are no connections between nodes in the same layer (in contrast to *unrestricted* Boltzmann machines, which permit connections between hidden units).

Reviewer: with one "visible" layer and ...

Reviewer 2: ```suggestion
**Overview**: RBMs are represented by undirected bipartite graphs, with one "visible" layer and a second "hidden" layer. There are no connections between nodes in the same layer (in contrast to *unrestricted* Boltzmann machines, which permit connections between hidden units).
```

**Learning**: We compute a loss function by marginalizing over all the hidden variables and taking the derivative w.r.t weights:

$$
\log L(v) = \log \sum_{h} \exp(\sum_{i,j} w_{ij} v_{i} h_{j} + \sum_{i} b_{i}v_{i} + \sum_{j} c_{j} h_{j} - \log(Z))
$$

$$
\frac{\partial \log L(v)}{\partial w_{ij}} = \sum_{h} P(h | v) \frac{\partial}{\partial w_{ij}} P(v,h) - \sum_{v,h} P(v,h) \frac{\partial}{\partial w_{ij}} P(v,h)
$$

or alternatively, as: 
$$
\dfrac{\partial}{\partial w_{ij}} \log L(v) = \mathbb{E}_{P(h|v)} \bigg[\dfrac{\partial}{\partial w_{ij}} P(v, h)\bigg] - \mathbb{E}_{P(v, h)} \bigg[\dfrac{\partial}{\partial w_{ij}} P(v, h)\bigg]
$$

Of the two terms in the loss above:

> - The first term is the expectation taking over the conditional of the hidden given the observed, ie. we *average over the posterior*. This is computed via sampling, which is exact (the RBM factorizes over $h$ given $v$).  In neural network literature, this is also called **the clamped**/ **wake**/ **positive phase**.

Reviewer: ie. -> i.e. 

what is "computed sampling"? please fix.

Author: ```suggestion
- The first term is the expectation taking over the conditional of the hidden given the observed, i.e. we *average over the posterior*. This is computed via sampling, which is exact (the RBM factorizes over $h$ given $v$).  In neural network literature, this is also called **the clamped**/ **wake**/ **positive phase**.
```

- The second term is the expectation taken over all random variables (both hidden and observed), i.e we *average over the joint*. This is also computed sampling, but requires the iterative Gibbs sampling algorithm. Consequently, it's far more expensive (computationally) and slow to converge. In neural network literature, this is called the **the unclamped**/ **sleep**/ **negative phase**.

## 2. Sigmoid Belief Networks

> SBNs consist of stacks of directed linear layers combined via non-linear sigmoid activation functions. The intermediate meaning of the hidden layers are typically ignored and so even though they synactically represent graphical models, operationally they begin to deviate. Operationally treating them as graphical models, leads to inefficient inference because of the implicit "explaining away" phenomenon as when we want to infer a hidden variable, it is coupled with all other hidden random variables in that layer. This stands in contrast to the RMB case because *d*-seperation in the RBM graph meant that the hidden random variables are independent given the other layer. Also, unlike the RBM, computing the conditional does not involve a negative phase.

Reviewer: RMB -> RBM

Reviewer 2: ```suggestion
SBNs consist of stacks of directed linear layers combined via non-linear sigmoid activation functions. The intermediate meaning of the hidden layers are typically ignored and so even though they synactically represent graphical models, operationally they begin to deviate. Operationally treating them as graphical models, leads to inefficient inference because of the implicit "explaining away" phenomenon as when we want to infer a hidden variable, it is coupled with all other hidden random variables in that layer. This stands in contrast to the RBM case because *d*-seperation in the RBM graph meant that the hidden random variables are independent given the other layer. Also, unlike the RBM, computing the conditional does not involve a negative phase.
```

We now connect our discussions of inference in the SBN and the RBM. Recall that for the gradient of the sleep phase of loss function, we could sample the joint using Gibbs procedure which alternates between different subsets of varaibles. While the vanilla Gibbs sample requires sampling every single random variable one by one given all other points, a broader idea, called block Gibbs sampling which groups the variables into arbitrary blocks and samples on the block given everything else. In the case of the RBM, we consider the two layers as two blocks and hence, sampling involves alternating between the two layers. 

<figure id="gibbs">

>       <img src="{{ 'assets/img/notes/lecture-15/gibbs.png' | relative_url }}" />

Reviewer: This figure is broken. Please fix.

Author: fixed

## 3. Deep Belief Networks

Independently of SBNs, people approached the problem of training deep architectures from a different lens and arrived at Deep Belief Nets. As discussed earlier, approaching the problem using exact inference is problematic due to the "explaining away" effect. Instead, Deep Belief Nets approach the problem as  hybrid graphical models which learn to extract deep hierarchical representations of the training data by stacking RBMs trained in a greedy manner followed by some ad-hoc fine-tuning.

More concretely, DBNs represent a joint probability distribution 

$$
P(v,h^1,h^2,h^3) = P(h^2,h^3) P(h^1|h^2) P(v|h^1) 
$$

where v are the visible variables and $ h^1 $, $ h^2 $ and $ h^3 $ are the hidden variables as shown in the figure below. Here, $ P(h^2,h^3) $ is an RBM and the conditionals $ P(h^1 \vert h^2) $ and $ P(v \vert h^1) $ are represented as sigmoid activations over a linear layer. As described earlier, directly performing inference and maximizing the likelihood of the joint is problematic. Hence we resort to a greedy layer-wise pretraining strategy described below:

> <figure id="DBN">

Reviewer: the figure is too big.. please make it smaller.

Author: fixed

>       <img width="50%" height="50%" src="{{ 'assets/img/notes/lecture-15/DBN.png' | relative_url }}" />

Author: ```suggestion
      <img width="216" height="276" src="{{ 'assets/img/notes/lecture-15/DBN.png' | relative_url }}" />
```

>       <img width="50%" height="50%" src="{{ 'assets/img/notes/lecture-15/DBN.png' | relative_url }}" />

Author: ```suggestion
      <img width="216" height="276" src="{{ 'assets/img/notes/lecture-15/DBN.png' | relative_url }}" />
```

  - **Iterate the previous two steps** : Repeat the procedure in previous two steps to train the desired number of layers by training successive RBMs using the hidden samples from the previous layer.
      Note that at any point, the last RBM (say the $N^{th}$ RBM) trained can be unrolled infinitely to produce an infinite belief network with the first N-1 layers given by the weights of the first N-1 trained RBMs and all the layers after that being represented using tied weights given by the $N^{th}$ RBM. Hence, each additional RBM trained can be viewed as untying an additional layer and freezing that layer of weights while fine-tuning all the following layers of the infinite belief network. Eg. In the figure below, the first 2 layers have been frozen and untied from the training procedure while all the layers starting from W3 are being trained as a standard RBM. 
        After pre-training the N RBMs, we simply stack them in the order in which they were trained to obtain an initialization of the DBN. 
        <figure id="Pre-training">

>     <img src="{{ 'assets/img/notes/lecture-15/pretraining.png' | relative_url }}" />

Reviewer: this figure is also way too big...

Reviewer 2: ```suggestion
    <img width="50%" height="50%" src="{{ 'assets/img/notes/lecture-15/pretraining.png' | relative_url }}" />
```

    - Pre-train a stack of RBMs in a greedy layer-wise fashion as in the autoencoder case.
    - Unroll the RBMs to create a feed-forward classifier. 
    - Fine-tune the parameters by backpropagating the classification error and optimizing by gradient descent.

<figure id="ae">
  <div class="row">
    <div class="col one">
      <img src="{{ 'assets/img/notes/lecture-15/ae-pretrain.png' | relative_url }}" />
    </div>
    <div class="col one">
      <img src="{{ 'assets/img/notes/lecture-15/ae-unroll.png' | relative_url }}" />
    </div>
  </div>
</figure>

## DBM: Deep Boltzmann Machines
<figure id="DBM">
      <img src="{{ 'assets/img/notes/lecture-15/DBM.png' | relative_url }}" />
</figure>
DBMs are a fully undirected extension of DBNs. They can be trained using a similar procedure as RBMs using MCMC or using a variational approximation of the data distribution and subsequently doing greedy layerwise pre-training as in the case of DBNs. As with DBNs these can also be used to initialize other networks for downstream tasks. 

> ## RBM v/s optimization steps

Reviewer: what does "v/s" mean?

Author: ```suggestion
## RBM v.s. optimization steps
```

$$
y^*(x,w) = \text{opt-alg}_y E(y,x;w)
$$

Here, $y^*$ is a non-linear differentiable function of the inputs and weights. Hence, we can impose some loss and optimize it as any other standard computation graph using backprop. Similarly, message passing based inference algorithms which can be truncated and converted into computational graphs and subsuquently optimizing the objective containing the approximate marginals (computed using the truncated message passing) directly to obtain better message passing and faster convergence.

> ## Concluding remarks on Graphical Models v/s Deep Networks

Reviewer: should be vs. ?

Reviewer 2: ```suggestion
## Concluding remarks on Graphical Models vs. Deep Networks
```
